javascript/1898-maximum-number-of-removable-characters.js

```diff
@@ -0,0 +1,96 @@
+/**
+ * https://leetcode.com/problems/maximum-number-of-removable-characters/
+ * 
+ * Brute force
+ * Time O(removable.length * s.length) | Space O(1)
+ * @param {string} s
+ * @param {string} p
+ * @param {number[]} removable
+ * @return {number}
+ */
+var maximumRemovals1 = function(s, p, removable) {
+  let k = 0;
+  // removable.reverse();
+  s = s.split('');
```

Comment:
This will increase space complexity

Reply:
I updated the space complexity to n. But we already had s and p as strings we just used them in an array instead of a string. wouldn't time complexity be the same for that? Before we were using n space for string which we were getting as an argument. Now we use the same number of array elements.

In other words, since we're just converting some string of n char to be an array of n elements. Then, the space complexity is the same as before right? because we were already using space(n) for the strings. Let me know if I'm making a mistake. Thank you.

Reply:
Sting is immutable primitive data type. You can not modify its state or space in memory.

If you split and convert to char array, it will need new space to allow the list to exist.

If you need to modify the string you can split, but if you need to read by index, you do not need to split and it will lower space complexity.

```js
/* Time O(N + N) | Space O(N)
for (const char of str.split('')) { ... }

/* Time O(N) | Space O(1)
for (const char of str) { ... }

/* Time O(N) | Space O(1)
for (const index char in str) {
	const char = str.charAt(index);
}
```

Reply:
@aakhtar3, I updated the space complexity.

Reply:
@aakhtar3, but if we don't convert it to the array then how are we going to update a specific char of the string? See line 18.
I did remove the split method on the binary-search solution and used the bitwise operation to find the mid now. Please check.